src/simularium/VisData.ts

```diff
@@ -369,36 +376,44 @@ class VisData {
      *   Functions to check update
      * */
     public hasLocalCacheForTime(timeNs: number): boolean {
+        const firstFrameTime = this.frameDataCache[0].time;
+        const lastFrameTime = this.frameDataCache[
+            this.frameDataCache.length - 1
+        ].time;
+
+        // Edge cases
         if (
             this.frameDataCache.length > 0 &&
             timeNs === 0 &&
-            this.frameDataCache[0].time <= Number.EPSILON // allow for floating point errors
+            firstFrameTime <= Number.EPSILON // allow for floating point errors
         ) {
+            // First frame is required and it exists in local cache
             return true;
         } else if (this.frameDataCache.length < 2) {
+            // Local cache only has 1 frame but we need something other than the first frame
             return false;
         }
 
-        return (
-            this.frameDataCache[0].time <= timeNs &&
-            this.frameDataCache[this.frameDataCache.length - 1].time >= timeNs
-        );
+        // Non-edge cases
+        const notLessThanFirstFrameTime =
+            compareFloats(timeNs, firstFrameTime, this.timeStepSize) !== -1;
+        const notGreaterThanLastFrameTime =
+            compareFloats(timeNs, lastFrameTime, this.timeStepSize) !== 1;
+        return notLessThanFirstFrameTime && notGreaterThanLastFrameTime;
     }
 
     public gotoTime(timeNs: number): void {
         this.cacheFrame = -1;
 
-        for (
-            let frame = 0, numFrames = this.frameDataCache.length;
-            frame < numFrames;
-            frame++
-        ) {
-            const frameTime = this.frameDataCache[frame].time;
-            if (timeNs <= frameTime) {
-                this.cacheFrame = Math.max(frame, 0);
-                break;
-            }
-        }
+        // Find the index of the frame that has the time matching our target time
+        const frameNumber = this.frameDataCache.findIndex((frameData) => {
+            return (
+                compareFloats(frameData.time, timeNs, this.timeStepSize) === 0
+            );
+        });
+
+        // frameNumber is -1 if findIndex() above doesn't find a match
+        this.cacheFrame = Math.max(frameNumber, 0);
```

Comment:
again, if the criteria is never met, do you want to still change this.cacheFrame?

Reply:
Fixed!